src/quark-core/src/QuarkNonceManager.sol

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.23;
+
+import {IQuarkWallet} from "quark-core/src/interfaces/IQuarkWallet.sol";
+
+/**
+ * @title Quark Nonce Manager
+ * @notice Contract for managing nonces for Quark wallets
+ * @author Compound Labs, Inc.
+ */
+contract QuarkNonceManager {
+    error NonReplayableNonce(address wallet, bytes32 nonce, bytes32 submissionToken);
+    error InvalidSubmissionToken(address wallet, bytes32 nonce, bytes32 submissionToken);
+
+    event NonceSubmitted(address wallet, bytes32 nonce, bytes32 submissionToken);
+
+    /// @notice Represents the unclaimed bytes32 value.
+    bytes32 public constant FREE = bytes32(uint256(0));
+
+    /// @notice A token that implies a Quark Operation is no longer replayable.
+    bytes32 public constant EXHAUSTED = bytes32(type(uint256).max);
+
+    /// @notice Mapping from nonces to last used submission token.
+    mapping(address wallet => mapping(bytes32 nonce => bytes32 lastToken)) public nonceSubmissions;
```

Comment:
`lastSubmissionToken`?

Reply:
this does break with the later comment that we should just settle on `NonceSubmission` as the only term we use... :wink: